File: evals/writing-tests-that-can-fail/recorded/2026-09-25/d-account-js/baseline-haiku-3/fixture_d_account_test.mjs
import test from 'node:test';
import assert from 'node:assert';
import { AccountRepo, Mailer, AccountService } from './fixture_d_account.mjs';

test('AccountService.deactivate', async (suite) => {
  const mockRepo = () => {
    const repo = Object.create(AccountRepo.prototype);
    repo.getCalls = [];
    repo.setStatusCalls = [];
    repo.get = function(userId) {
      this.getCalls.push(userId);
      return { status: 'active', email: 'user@example.com' };
    };
    repo.setStatus = function(userId, status) {
      this.setStatusCalls.push({ userId, status });
    };
    return repo;
  };

  const mockMailer = () => {
    const mailer = Object.create(Mailer.prototype);
    mailer.sendGoodbyeCalls = [];
    mailer.sendGoodbye = function(email) {
      this.sendGoodbyeCalls.push(email);
    };
    return mailer;
  };

  await suite.test('deactivates an active account', () => {
    const repo = mockRepo();
    const mailer = mockMailer();
    const service = new AccountService(repo, mailer);

    service.deactivate('user123');

    assert.deepEqual(repo.getCalls, ['user123']);
    assert.deepEqual(repo.setStatusCalls, [{ userId: 'user123', status: 'inactive' }]);
    assert.deepEqual(mailer.sendGoodbyeCalls, ['user@example.com']);
  });

  await suite.test('skips operations when account is already inactive', () => {
    const repo = mockRepo();
    repo.get = function(userId) {
      this.getCalls.push(userId);
      return { status: 'inactive', email: 'user@example.com' };
    };
    const mailer = mockMailer();
    const service = new AccountService(repo, mailer);

    service.deactivate('user123');

    assert.deepEqual(repo.getCalls, ['user123']);
    assert.deepEqual(repo.setStatusCalls, []);
    assert.deepEqual(mailer.sendGoodbyeCalls, []);
  });

  await suite.test('retrieves account by correct userId', () => {
    const repo = mockRepo();
    const mailer = mockMailer();
    const service = new AccountService(repo, mailer);

    service.deactivate('specific-user-id');

    assert.strictEqual(repo.getCalls[0], 'specific-user-id');
  });

  await suite.test('sends goodbye email to correct email address', () => {
    const repo = mockRepo();
    repo.get = function(userId) {
      this.getCalls.push(userId);
      return { status: 'active', email: 'john@example.com' };
    };
    const mailer = mockMailer();
    const service = new AccountService(repo, mailer);

    service.deactivate('user123');

    assert.strictEqual(mailer.sendGoodbyeCalls[0], 'john@example.com');
  });

  await suite.test('sets status to inactive on repo', () => {
    const repo = mockRepo();
    const mailer = mockMailer();
    const service = new AccountService(repo, mailer);

    service.deactivate('user456');

    assert.strictEqual(repo.setStatusCalls[0].status, 'inactive');
  });

  await suite.test('passes correct userId to setStatus', () => {
    const repo = mockRepo();
    const mailer = mockMailer();
    const service = new AccountService(repo, mailer);

    service.deactivate('target-user');

    assert.strictEqual(repo.setStatusCalls[0].userId, 'target-user');
  });

  await suite.test('returns undefined', () => {
    const repo = mockRepo();
    const mailer = mockMailer();
    const service = new AccountService(repo, mailer);

    const result = service.deactivate('user123');

    assert.strictEqual(result, undefined);
  });

  await suite.test('handles multiple deactivation calls', () => {
    const repo = mockRepo();
    const mailer = mockMailer();
    const service = new AccountService(repo, mailer);

    service.deactivate('user1');
    service.deactivate('user2');

    assert.deepEqual(repo.getCalls, ['user1', 'user2']);
    assert.strictEqual(repo.setStatusCalls.length, 2);
    assert.strictEqual(mailer.sendGoodbyeCalls.length, 2);
  });

  await suite.test('handles account with different statuses', () => {
    const repo = mockRepo();
    repo.get = function(userId) {
      this.getCalls.push(userId);
      return { status: 'suspended', email: 'user@example.com' };
    };
    const mailer = mockMailer();
    const service = new AccountService(repo, mailer);

    service.deactivate('user123');

    assert.deepEqual(repo.setStatusCalls, [{ userId: 'user123', status: 'inactive' }]);
    assert.deepEqual(mailer.sendGoodbyeCalls, ['user@example.com']);
  });
});

File: evals/writing-tests-that-can-fail/recorded/2026-09-25/d-account-py/baseline-sonnet-1/fixture_d_account_test.py
"""Tests for fixture_d_account.py."""

import pytest

from fixture_d_account import Account, AccountRepo, AccountService, Mailer


class FakeAccountRepo(AccountRepo):
    def __init__(self, account: Account) -> None:
        self.account = account
        self.set_status_calls = []

    def get(self, user_id: str) -> Account:
        assert user_id == self.account.user_id
        return self.account

    def set_status(self, user_id: str, status: str) -> None:
        self.set_status_calls.append((user_id, status))
        self.account.status = status


class FakeMailer(Mailer):
    def __init__(self) -> None:
        self.sent_to = []

    def send_goodbye(self, email: str) -> None:
        self.sent_to.append(email)


@pytest.fixture
def active_account() -> Account:
    return Account(user_id="u1", email="u1@example.com", status="active")


@pytest.fixture
def inactive_account() -> Account:
    return Account(user_id="u2", email="u2@example.com", status="inactive")


def test_deactivate_active_account_sets_status_and_sends_email(active_account):
    repo = FakeAccountRepo(active_account)
    mailer = FakeMailer()
    service = AccountService(repo, mailer)

    service.deactivate("u1")

    assert repo.set_status_calls == [("u1", "inactive")]
    assert mailer.sent_to == ["u1@example.com"]
    assert active_account.status == "inactive"


def test_deactivate_already_inactive_account_is_noop(inactive_account):
    repo = FakeAccountRepo(inactive_account)
    mailer = FakeMailer()
    service = AccountService(repo, mailer)

    service.deactivate("u2")

    assert repo.set_status_calls == []
    assert mailer.sent_to == []


def test_deactivate_looks_up_correct_user_id(active_account):
    repo = FakeAccountRepo(active_account)
    mailer = FakeMailer()
    service = AccountService(repo, mailer)

    with pytest.raises(AssertionError):
        service.deactivate("someone-else")


def test_account_is_dataclass_with_expected_fields():
    account = Account(user_id="u3", email="u3@example.com", status="active")

    assert account.user_id == "u3"
    assert account.email == "u3@example.com"
    assert account.status == "active"


def test_account_repo_base_methods_raise_not_implemented():
    repo = AccountRepo()

    with pytest.raises(NotImplementedError):
        repo.get("u1")

    with pytest.raises(NotImplementedError):
        repo.set_status("u1", "inactive")


def test_mailer_base_method_raises_not_implemented():
    mailer = Mailer()

    with pytest.raises(NotImplementedError):
        mailer.send_goodbye("u1@example.com")
